README.md

```diff
@@ -1,18 +1,129 @@
-extensible-effects is based on the work
-[Extensible Effects: An Alternative to Monad Transformers](http://okmij.org/ftp/Haskell/extensible/).
-Please read the [paper](http://okmij.org/ftp/Haskell/extensible/exteff.pdf) and
-the followup [freer paper](http://okmij.org/ftp/Haskell/extensible/more.pdf) for
-details. Additional explanation behind the approach can be found on [Oleg's website](http://okmij.org/ftp/Haskell/extensible/).
+
+# Extensible effects
 
 [![Build Status](https://travis-ci.org/suhailshergill/extensible-effects.svg?branch=master)](https://travis-ci.org/suhailshergill/extensible-effects)
 [![Join the chat at https://gitter.im/suhailshergill/extensible-effects](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/suhailshergill/extensible-effects?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
 [![Stories in Ready](https://badge.waffle.io/suhailshergill/extensible-effects.png?label=ready&title=Ready)](http://waffle.io/suhailshergill/extensible-effects)
 [![Stories in progress](https://badge.waffle.io/suhailshergill/extensible-effects.png?label=in%20progress&title=In%20progress)](http://waffle.io/suhailshergill/extensible-effects)
 
-## Advantages
+*Implement effectful computation in a modular way!*
+
+The main and only monad is built upon `Eff` from `Control.Eff`. `Eff r a` is parameterized by
+the effect-list `r` and the monadic-result type `a` similar to other monads.
+
+In case you know monad transformers or mtl: This library provides only one monad that includes all your effects instead of layering different transformers. It is not necessary to lift the computations through a monad stack. Also, it is not required to lift every `Monad*` (like `MonadError`) typeclass though all transformers.
+
+## Quickstart
+
+### The effect list
+
+The effect list `r` in the type `Eff r a` is a central concept in this library.
+It is a type-level list containing effect types. If `r` is ... then,
+
+If `r` is the empty list, then the computation `Eff r` (or `Eff '[]`) does not contain any effects to be handled and therefore is a pure computation. In this case, the result value can be retrieved by `run :: Eff '[] a -> a`
+
+For programming within the `Eff r` monad, it is never necessary to list all effects that can appear.
+It suffices to state what types of effects are at least required.
+This is done via the `Member t r` typeclass. It describes that the type `t` occurs inside the list `r`.
+
+Effect-handler functions, mostly called `run*`, `exec*` or `eval*`, typically have a signature like `Eff (x ': r) a -> Eff r a`. Where `x` is the type of effect to be handled, like `Writer w` or `State s`. The transformation from the longer list of effects `(x ': r)` to just `r` is a type-level indicator that the effect `x` has been handled. Depending on the effect, some additional input might be required or some output is produced.
```

Comment:
@power-fungus additionally, introducing effects are done via an EDSL (embedded DSL) that we custom make for each effect for each effect. the handler of the effect, is the interpreter of this EDSL.

the general pattern for the EDSL (in this library) is to have a GADT encapsulating the algebra. each operation in the EDSL (i.e., constructor for the GADT) also has an equivalent function which encapsulates the request to the handler of the effect. and finally, there is the interpreter, i.e., the `handler` function. 

in the spirit of extensibility, not only can we have custom effects, but for each effect we can have multiple handlers (aka interpreters) and choose which one to leverage at the call-site.

Reply:
@suhailshergill I explicitly wanted to write `Eff r a`, since the `a` type is still found *somewhere* in the resulting type in virtually all cases. And it was my intent to hint at that fact. Of course, this is not *correct* but I think gives good intuition. 

Also, I would not mention the word "EDSL" in the introduction. Haskell programmers know what functions are, and for this audience, knowledge of monads is also expected. My goal is that this is totally enough to get started. In my opinion, the additional layer is not too helpful.

I see that the dual concepts of requesting and handling are quite important. I try to explain this in the next iteration, while still getting to the first examples fairly quickly.

Contrary to most Haskell documentation, I try to be not too technically correct since such documentation is readable, only if you already know the library well already. I want the readers to build intuition, which serves as a starting point for the documentations of the single modules, which should be 100% correct.

Reply:
@power-fungus excellent points. i agree on the importance of optimising for comprehension. that being said, would it make sense for some of this to be moved to the "advanced users" section?

Reply:
@suhailshergill There will probably no dedicated advanced section. There will be the module-documentations and there will be a description of how to implement a custom effect. I guess that this is the more fitting.